msal_extensions/osx.py

```diff
@@ -0,0 +1,315 @@
+# pylint: disable=duplicate-code
+
+"""Implements a macOS specific TokenCache, and provides auxiliary helper types."""
+
+import os
+import ctypes as _ctypes
+
+OS_RESULT = _ctypes.c_int32
+
+
+class KeychainError(OSError):
+    """The parent of all exceptions that may be thrown by the functions implemented inside of the
+    Keychain type."""
+    def __init__(self, exit_status):
+        super(KeychainError, self).__init__()
+        self.exit_status = exit_status
+        # TODO: pylint: disable=fixme
+        #  use SecCopyErrorMessageString to fetch the appropriate message here.
+        self.message = \
+            '{} ' \
+            'see https://opensource.apple.com/source/CarbonHeaders/CarbonHeaders-18.1/MacErrors.h'\
+                .format(self.exit_status)
+
+
+class KeychainAccessDeniedError(KeychainError):
+    """The exceptions that's raised when a Keychain is set to be queried, but the current process
+    has not been give permission to access it, and a query comes anyway."""
+    EXIT_STATUS = -128
+
+    def __init__(self):
+        super(KeychainAccessDeniedError, self).__init__(KeychainAccessDeniedError.EXIT_STATUS)
+
+
+class NoSuchKeychainError(KeychainError):
+    """The exception that's raised when a Keychain is set to be queried, but it is non-existent, and
+    a query comes anyway."""
+    EXIT_STATUS = -25294
+
+    def __init__(self, name):
+        super(NoSuchKeychainError, self).__init__(NoSuchKeychainError.EXIT_STATUS)
+        self.name = name
+
+
+class NoDefaultKeychainError(KeychainError):
+    """The exception that's raised when no Keychain is set to be queried, but a query comes
+    anyway.
+    """
+    EXIT_STATUS = -25307
+
+    def __init__(self):
+        super(NoDefaultKeychainError, self).__init__(NoDefaultKeychainError.EXIT_STATUS)
+
+
+class KeychainItemNotFoundError(KeychainError):
+    """The exception that's raised when a non-exist Keychain entry is requested."""
+    EXIT_STATUS = -25300
+
+    def __init__(self, service_name, account_name):
+        super(KeychainItemNotFoundError, self).__init__(KeychainItemNotFoundError.EXIT_STATUS)
+        self.service_name = service_name
+        self.account_name = account_name
+
+
+def _construct_error(exit_status, **kwargs):
+    if exit_status == KeychainAccessDeniedError.EXIT_STATUS:
+        return KeychainAccessDeniedError()
+    if exit_status == NoSuchKeychainError.EXIT_STATUS:
+        return NoSuchKeychainError(**kwargs)
+    if exit_status == NoDefaultKeychainError.EXIT_STATUS:
+        return NoDefaultKeychainError()
+    if exit_status == KeychainItemNotFoundError.EXIT_STATUS:
+        return KeychainItemNotFoundError(**kwargs)
+    return KeychainError(exit_status)
```

Comment:
Syntactic sugar:

```python
def _construct_error(exit_status, **kwargs):
    return {
        KeychainAccessDeniedError.EXIT_STATUS: KeychainAccessDeniedError,
        ...
        }.get(exit_status, KeychainError)(exit_status, **kwargs)
```

Reply:
That looks a lot better! And honestly given the argument above, that this library is the only one that will ever actually construct one of these errors, I think it is enough prettier that it warrants the misleading nature of the additional parameter.

Reply:
Fixed in 2962bd8d69a24544a0f6f725e7e0e89cb16fbcea

Reply:
Resolving this conversation after [offline discussion](https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/15#issuecomment-505079762)